src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+package org.rascalmpl.repl;
+
+import java.io.FilterOutputStream;
+import java.io.FilterWriter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.nio.charset.Charset;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.rascalmpl.debug.IRascalMonitor;
+import io.usethesource.vallang.ISourceLocation;
+import jline.Terminal;
+
+/**
+ * The terminal progress bar monitor wraps the standard output stream to be able to monitor
+ * output and keep the progress bars at the same place in the window while other prints happen
+ * asyncronously (or even in parallel). It can be passed to the IDEServices API such that
+ * clients can start progress bars, make them grow and end them using the API in util::Monitor.
+ * 
+ * This gives the console the ability to show progress almost as clearly as an IDE can with a
+ * UI experience. 
+ * 
+ * This class only works correctly if the actual _raw_ output stream of the terminal is wrapped
+ * with an object of this class.
+ */
+public class TerminalProgressBarMonitor extends FilterOutputStream implements IRascalMonitor  {
+    /**
+     * We administrate an ordered list of named bars, which will be printed from
+     * top to bottom just above the next prompt.
+     */
+    private List<ProgressBar> bars = new LinkedList<>();
```

Comment:
can this not be final? why does it have to be a linked list? in most cases arraylist are nicer, as they have better locality.